### PR TITLE
Let expandable segments reuse free blocks larger than max_split_size

### LINKED ISSUE #123548: An error occurs when ‘max_split_size_mb ’and ‘expandable_segments ’ are enabled at the same time.
### 🐛 Describe the bug

```python
import os
os.environ["PYTORCH_CUDA_ALLOC_CONF"]='max_split_size_mb:500,expandable_segments:True'
import torch

def create_tensor():
    x = torch.zeros([50,800,1200,3], device='cuda:0')  # 550M     free_block1

if __name__ == '__main__':
    create_tensor()
    temp = []
    number = 500  # Adjusted based on the CUDA memory size.
    for _ in range(number):
        x=torch.zeros([10,800,1200,3], device='cuda:0') # 110MB block2
        temp.append(x)

```
`block2` is continuously allocated and allocated memory close to the upper limit triggers oom. The `release_available_cached_blocks` function is called.But using release_block in this function to release the memory is problematic.
https://github.com/pytorch/pytorch/blob/26bf05ccacc0377f0ef40d1d9c792c403267d5d5/c10/cuda/CUDACachingAllocator.cpp#L2470
https://github.com/pytorch/pytorch/blob/26bf05ccacc0377f0ef40d1d9c792c403267d5d5/c10/cuda/CUDACachingAllocator.cpp#L2500
In `release_block`, the `cudaFree` function is used to release the memory. The correct method is to use the `unmap` function.

In addition, I did not find in the documentation that the two environment variables cannot be shared.

### Versions

PyTorch version: main branch

### PR DESCRIPTION
**Pull request by doplxyz**
Fixes #123548.

### The problem

With `max_split_size_mb` and `expandable_segments:True` set together, `reserved` memory
grows without bound while the live set stays small, until the device is full and the
allocator has to release and retry.

Reproducer using nothing but `torch.empty`, on a 16 GiB card:

```
PYTORCH_ALLOC_CONF=max_split_size_mb:128,expandable_segments:True
```

| cycle | 0 | 20 | 40 | 60 | 80 | 100 |
|---|---|---|---|---|---|---|
| `allocated` (MiB) | 66 | 106 | 146 | 186 | 226 | 266 |
| `reserved` (MiB) | 220 | 2,980 | 5,760 | 8,520 | 11,300 | 14,060 |

Each cycle allocates A = 91 MiB and B = 45.5 MiB, then a 2 MiB block that stays alive,
then frees A and B. Either option on its own, or the same allocation pattern without the
small block that stays alive, keeps `reserved` under 500 MiB.

### Why it happens

`get_free_block()` applies the `max_split_size` oversize rules to every candidate block,
including blocks that live inside an expandable segment:

```c++
    // Do not return an oversized block for a large request
    if ((p.size() < AcceleratorAllocatorConfig::max_split_size()) &&
        ((*it)->size >= AcceleratorAllocatorConfig::max_split_size()))
      return false;
```

In the reproducer, freeing A and B leaves a merged 136.5 MiB free block. The 2 MiB block
allocated after them is still alive, so that merged block sits in the interior of the
segment and is not adjacent to its unmapped tail. The next 91 MiB request is smaller than
`max_split_size` while the block is larger than it, so the block is rejected, the
expandable-segment path cannot reach it either, and the segment is grown instead.

These rules exist because splitting a standalone `cudaMalloc`ed block keeps the whole
block pinned until every piece of it is free again. That does not apply inside an
expandable segment, where pages are mapped and unmapped independently: handing out part of
a large free block there does not pin the rest of the segment. So the rules reject memory
that is in fact reusable, and force the segment to grow instead.

### The change

Both oversize checks in `get_free_block()` are skipped for blocks that belong to an
expandable segment.

The condition tests the *block* (`(*it)->expandable_segment_`), not the current setting
(`p.is_expandable_segments_active`). Testing the setting would also lift the rules for
ordinary blocks that were created before expandable segments were switched on, which is
not intended — the reason the rules do not apply is a property of the block, not of the
configuration in effect at the time of the request.

This makes `get_free_block()` consistent with the two other places in the same file that
already treat expandable segments as an exception:

- `should_split()` ignores `max_split_size` when expandable segments are active
- `release_available_cached_blocks()` does not release expandable segments (#134338)

`get_free_block()` was the remaining one.

### Test

`test_max_split_expandable_reuses_interior_blocks` in `test/test_cuda.py` runs the same
pattern at reduced scale. It is parametrized over the two oversize rules the change
touches, since sizes decide which one a request hits:

| case | request | merged free block | rule reached |
|---|---|---|---|
| `big_request=False` | 24 MiB (< `max_split_size` 32 MiB) | 36 MiB (>= 32 MiB) | first |
| `big_request=True` | 64 MiB (>= `max_split_size`) | 96 MiB (>= 64 + 20 MiB) | second |

`max_non_split_rounding_mb` is set explicitly by the test rather than left at its default,
so a change to that default cannot silently stop the second case from reaching the rule it
targets. The assertions are derived from the mechanism rather than a fixed number: no
single cycle may grow `reserved` by as much as the merged block, and the cycles combined
may not grow by one merged block either. The test also asserts that an expandable segment
exists before it measures, so it cannot pass by falling back to the ordinary allocator.
It peaks at a few hundred MiB, so it does not need a large card.

### Before/after

Measured on two clean local builds of this PR's parent (`04de8a6b2a`), ROCm 7.2.4 /
gfx1201, differing only in this PR's allocator change:

| | before | after |
|---|---|---|
| `..._big_request_False` | FAILED | PASSED |
| `..._big_request_True` | FAILED | PASSED |
| `TestCudaAllocator -k "expandable or max_split"` | 3 passed, 1 skipped | 5 passed, 1 skipped |
| reserved growth, first rule, 5 cycles | 200 MiB | 20 MiB |
| reserved growth, second rule, 5 cycles | 500 MiB | 20 MiB |

An auxiliary measurement using the same allocation sequence also records addresses: before
the change, the request landed inside the free range it should have reused in 0 of 5
cycles; after, in 5 of 5, with the whole requested interval contained in that range. The
full provenance of the two builds and the raw numbers are in a comment below.

### Verification status

Verified locally only on ROCm 7.2.4 / gfx1201 (Radeon RX 9070 XT, single GPU). The edit is
in `c10/cuda/CUDACachingAllocator.cpp`; what ran on this machine is the hipified copy of
the same conditional, so the source edit is shared but the generated code and the runtime
underneath are not. I do not have a CUDA device, so the CUDA path is unverified locally
and I am relying on CI for it.

CI has not run on this PR yet: the workflow runs are sitting at `action_required`, which I
believe needs a maintainer to approve them for a first-time contributor. I would
appreciate that approval — the CUDA side of this change is exactly what I cannot check
myself.

### Notes

I found this while chasing a ROCm-specific problem where the resulting map/unmap churn
also loses the contents of a live allocation (ROCm/ROCm#6603). That part is not addressed
here and is not what this PR claims to fix. The evidence in this PR covers block selection
in `get_free_block()` and the resulting reserved-memory growth, and says nothing about
data integrity.

### CODE REVIEW

**Comment by doplxyz** [comment]
### Before/after from a local build, and the test now covers both oversize rules

I have the after measurement that was missing when I opened this, and I have extended the
test to cover the second oversize rule as well.

#### How the two builds were produced

Both are clean full builds — `rm -rf build` first, not incremental — of one worktree in
one container, driven by the same script.

The worktree stays checked out at this PR's parent, `04de8a6b2a0383d8725a97386fd96060b9e28dfa`,
for both builds. The test file from PR head `a1705e9ba8a02c69f8decaf4f5a7c3861a14c425` is
present in both, so the before run exercises exactly the same assertions. For the after
build only, that head's `c10/cuda/CUDACachingAllocator.cpp` is applied on top as an
uncommitted change. That is why `torch.version.git_version` reports the parent SHA in both
processes.

```
worktree at 04de8a6b2a0383d8725a97386fd96060b9e28dfa
container rocm/pytorch:rocm7.2.4_ubuntu24.04_py3.12_pytorch_release_2.9.1
ROCM_PATH=/opt/rocm-7.2.4 PYTORCH_ROCM_ARCH=gfx1201 USE_ROCM=1
python tools/amd_build/build_amd.py     # hipify, required before configure
python setup.py develop
```

The builds are told apart by the binaries themselves:

| | before | after |
|---|---|---|
| `libc10_hip.so` build-id | `b0eb92a677f8d8fd3b773a3aec130492af1094a8` | `fa37813b68ae6db8d09eeba4d7a3161f1a7a6d30` |
| `if (!(*it)->expandable_segment_)` in `c10/hip/HIPCachingAllocator.cpp` | absent | present |

The ROCm build compiles a hipified copy of `CUDACachingAllocator.cpp`, so
`c10/hip/HIPCachingAllocator.cpp` is the file that actually gets compiled here.

The paths below come from `/proc/self/maps`, read after PyTorch initialized the ROCm
device runtime, so they are the DSOs the process really mapped rather than where the
Python package happens to live. The build-ids are then read from those exact paths — the
maps file itself does not carry them. Only one copy of each name was mapped in any run.

```
libc10_hip.so    /opt/venv/lib/python3.12/site-packages/torch/lib/libc10_hip.so
libtorch_hip.so  /opt/venv/lib/python3.12/site-packages/torch/lib/libtorch_hip.so
libamdhip64.so   /opt/rocm-7.2.4/lib/libamdhip64.so.7.2.70204
```

One deviation from the stock container: `torchvision` and `torchaudio` ship in that image
built against its bundled torch 2.9.1, and their extensions fail to load against a
locally built torch (`operator torchvision::nms does not exist`) which breaks collection
of `test_cuda.py`. I uninstalled both. They are not involved in the allocator path.

#### The interior block is actually reused

To be clear about what is committed and what is not: **the regression test asserts the
reserved-memory bounds and the presence of an expandable segment. It does not assert
anything about addresses.** The address evidence below comes from a separate auxiliary
script that runs the same allocation sequence outside the test suite.

Reserved-memory totals alone do not show *which* block a request came from. The auxiliary
script therefore records addresses. Each cycle notes the half-open range spanned by A and
B before freeing them,

```
merged = [min(A, B), max(A + sizeA, B + sizeB))
```

and checks that this pair really is one that merges on free. Adjacency alone is not
enough — two adjacent addresses in different segments would not merge — so each pair is
checked for three things: `max - min == sizeA + sizeB`, both blocks in the same segment,
and that segment reporting `is_expandable: true`. This is recorded for every pair, the
warm-up one and all five measured ones, and all three held for 6 of 6 pairs in each of the
six processes.

The next cycle then checks its own request against the previous range. Both the start
address and the whole requested interval `[p, p + request)` are checked; the numbers below
are identical either way, so the containment is not a coincidental overlap at the head of
the range.

Settings are `expandable_segments:True,max_split_size_mb:32,max_non_split_rounding_mb:20`,
one warm-up cycle, then five consecutive measured cycles in a single process. This is one
stateful sequence, not five independent measurements; it was repeated in three fresh
processes per build, six in total, which agreed exactly. In every one of them exactly one
path was mapped per DSO name, so there is no second copy of `libc10_hip.so` in play.

| A + B (merged) | request that must reuse it | rule | landed in previous merged range | reserved growth per cycle | total |
|---|---|---|---|---|---:|
| **before** | | | | | |
| 24 + 12 MiB (36) | 24 MiB | first | **0 of 5** | 40, 40, 40, 40, 40 | 200 MiB |
| 64 + 32 MiB (96) | 64 MiB | second | **0 of 5** | 100, 100, 100, 100, 100 | 500 MiB |
| **after** | | | | | |
| 24 + 12 MiB (36) | 24 MiB | first | **5 of 5** | 0, 20, 0, 0, 0 | 20 MiB |
| 64 + 32 MiB (96) | 64 MiB | second | **5 of 5** | 0, 20, 0, 0, 0 | 20 MiB |

In every before cycle the request landed outside the range it should have reused, and the
segment grew instead. The growth per cycle is the merged block plus the 2 MiB fence,
rounded up to the expandable segment's mapping granule: `round_up(36 + 2, 20) = 40` and
`round_up(96 + 2, 20) = 100`, which matches the measured values exactly. (Two unrelated
quantities happen to both be 20 MiB here: this mapping granule, and the
`max_non_split_rounding_mb` the test sets. The setting does not determine the granule.)

After the change every request landed inside the previous merged range, and four of the
five cycles add no reserved memory at all. The single 20 MiB step is not what establishes
reuse — the address containment above is. It is consistent with one additional mapping
granule as the 2 MiB fences accumulate (10 MiB over the five measured cycles; the warm-up
fence is allocated before the baseline is taken), but I did not trace the mapping calls to
confirm that attribution.

The `is_expandable: true` check above covers every measured cycle, not just the first, so
the ordinary allocator was not silently substituted partway through.

#### Both oversize rules are exercised

The first version of the test only reached the first rule. `get_free_block()` has two and
this PR changes both, so the test is parametrized to construct a case for each:

- `request < max_split_size` and `block >= max_split_size` — first rule
- `request >= max_split_size` and `block >= request + max_non_split_rounding_size` —
  second rule

`max_non_split_rounding_mb` is now set explicitly by the test rather than left at its
default, so a change to that default cannot silently stop the second case from reaching
the rule it targets. The assertions are derived from the mechanism rather than a fixed
number: no single cycle may grow reserved by as much as the merged block, and the cycles
combined may not grow by one merged block either. The test also asserts that an expandable
segment exists before it measures.

#### Test results

| | before | after |
|---|---|---|
| `..._reuses_interior_blocks_big_request_False` | FAILED | PASSED |
| `..._reuses_interior_blocks_big_request_True` | FAILED | PASSED |
| `TestCudaAllocator -k "expandable or max_split"` | 3 passed, 1 skipped, 2 failed (the two above) | 5 passed, 1 skipped |

The skip in both columns is `test_max_split_expandable`, disabled on Linux by #180263, so
it is not part of the evidence here. The command was:

```
python -m pytest -v test/test_cuda.py::TestCudaAllocator -k "expandable or max_split"
```

#### What this does not establish

- This was run only on ROCm 7.2.4 / gfx1201 (Radeon RX 9070 XT, 16 GiB, single GPU). I do
  not have a CUDA device, so the CUDA path is unverified locally and I am relying on CI
  for it. The edit is in `c10/cuda/CUDACachingAllocator.cpp`, and what ran here is the
  hipified copy of that same conditional — the source edit is shared, the generated code
  and the runtime underneath are not.
- Both builds are local builds of this PR's parent, not CI builds.
- The scope of this evidence is block selection in `get_free_block()` and the resulting
  reserved-memory growth. It says nothing about data integrity.

I can run anything else on this card if a different configuration would be more
convincing.

**Comment by doplxyz** [comment]
New measurements bearing on this PR, plus one caveat about what they do and do not show.

I ran the reproducer from the PR description across six builds, holding everything else fixed:
same base image (`python:3.12-slim-bookworm@sha256:a116514e…` plus a fixed set of OS libraries),
same script, same `PYTORCH_ALLOC_CONF=max_split_size_mb:128,expandable_segments:True`, same GPU
(RX 9070 XT, gfx1201, 16,304 MiB, sole occupant), same host. Ten runs per build, one round at a
time in rotation, each run in a fresh container, each waiting for VRAM to be released first.

| torch | bundled ROCm | peak `reserved` |
|---|---|---:|
| 2.9.1 | 7.14 | **782 MiB** |
| 2.12.0 | 7.2 | 16,000 MiB |
| 2.12.1 | 7.2 | 16,000 MiB |
| 2.13.0 | 7.2 | 16,000 MiB |
| 2.12.0 | 7.14 | 15,860–16,000 MiB |
| 2.14.0.dev20260811 | 7.14 | 15,860–16,000 MiB |

**The growth is not specific to one ROCm userspace.** It reaches the same 16,000 MiB ceiling on a
ROCm 7.2 stack and on a 7.14 one. I have only tested gfx1201, so I am not claiming anything about
other architectures, and I have not excluded the runtime from playing some role in the timing —
but it is clearly not confined to one of the two stacks.

**2.9.1 does not exhibit the growth** (782 MiB). That is a
statement about `reserved` only; I am not drawing any conclusion from it about the ROCm-side
issue below. These builds differ in
more than the PyTorch version — bundled runtime and dependencies differ too — so this is not a
clean bisect. It bounds where the behaviour appears rather than attributing it to a specific
commit range; I have not bisected.

I also varied only the allocator options, on one build (torch 2.12.0, ROCm 7.14), n=10 per row.
The merged free block the reproducer creates is 136.5 MiB:

| `PYTORCH_ALLOC_CONF` | peak `reserved` |
|---|---:|
| `expandable_segments:True` | 780 MiB |
| `max_split_size_mb:128` | 782 MiB |
| `max_split_size_mb:256,expandable_segments:True` | 780 MiB |
| `max_split_size_mb:128,expandable_segments:True` | **16,000 MiB** |

This is consistent with the mechanism proposed in the PR body — the growth appeared at 128,
below the merged free-block size, and not at 256, above it — but I have not instrumented
`get_free_block()` or the expandable-segment path, so I am reporting the outcome rather than
confirming which branch was taken. I also tested only these two values, on this reproducer.

On all three ROCm 7.2 builds the run reaches 16,000 MiB of the 16,304 MiB on the card, hits its
first allocation failure between cycle 100 and 120, recovers, and continues to cycle 280 with
`num_alloc_retries` at 56 and `num_ooms` at 0, staying bit-exact throughout. So on that stack the growth is a
memory-efficiency problem, as described here — the allocator recovers by releasing and retrying.

There is a second reason to bound it, which I did not know when I opened this PR. On a ROCm 7.14
userspace, in the same configuration, a live allocation that user code never writes to reads back
zeroed from offset 0 — 40 to 64 MiB of a 64 MiB buffer, depending on the run — with no exception
and nothing in `journalctl -k`. That happened in 19 of 20 runs across
torch 2.12.0 and 2.14.0.dev, and in 0 of 30 runs on ROCm 7.2 with 2.12.0 / 2.12.1 / 2.13.0. I have
filed it separately as ROCm/ROCm#6603, and I am not claiming it is caused by this allocator
behaviour — the 7.2 builds reach `max_reserved` of up to 16,000 MiB and stay correct, so some difference between the
tested artifacts may be involved as well. Those artifacts differ in more than the ROCm userspace, so I
am not attributing it to a component.

What the ablation above does show is that across those four configurations on the affected stack,
the growth and the corruption co-occur exactly: the three configurations that did not produce the
growth did not corrupt in 10 runs each, and the one that did corrupted in 10 of 10. **I have not
tested this PR's patch on a ROCm 7.14 build, and I am not claiming it changes that behaviour.**
The ROCm-side fault is unlocalised, and I have not shown what the causal relationship between the
growth and the zeroing is — only that in this reproducer they appear and disappear together.

The per-run logs, the classification script and the per-cell environment manifests for all of the
above are at https://github.com/doplxyz/rocm6603-vmm-canary-evidence.

Happy to rerun any of this, fold the cross-version numbers into the PR description, build and test
the patched allocator on the 7.14 stack, or extend the test in `test/test_cuda.py` — whichever is
most useful for review.
